### PR TITLE
Leds?

### PR DESCRIPTION
**Pull request by BenBerol**
So I don't know why it shows all these file conflicts when I make the pull request, because a lot of the differences it is noticing are actually the same as the version of your code that I see from your commit 7 hours ago. However, when I try and rebase it gives me a bunch of merge conflicts that don't end up solving the problem. I went through them all once, got a bunch of errors, fixed them, and then when I pushed came back to make the pull request I had more changed files then before. Is there a way I can just force in all of the changes, because I think they work and are basically the same as your most up to date commit

### CODE REVIEW

**Comment by BenBerol** [comment]
Yay! It looks like I was able to git rid of the merge conflicts by manually merging on github (ew gross). Let me know if there's something else I need to change!

File: PathWeaver/Autos/charge [hunk 1]
```diff
@@ -0,0 +1 @@
+charge.path
```
Comment (BrandonWerbel) [inline PathWeaver/Autos/charge]: Delete this file

File: src/main/deploy/output/charge.wpilib.json [hunk 2]
```diff
@@ -0,0 +1,1817 @@
+[
```
Comment (BrandonWerbel) [inline src/main/deploy/output/charge.wpilib.json]: Delete this file

File: src/main/java/frc/robot/Robot.java [hunk 3]
```diff
@@ -38,6 +38,7 @@ public void robotInit() {
     // Instantiate our RobotContainer.  This will perform all our button bindings, and put our
     // autonomous chooser on the dashboard.
     m_robotContainer = new RobotContainer(this);
+    m_robotContainer.startLEDs();
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/Robot.java]: Can you do this in the RobotContainer constructor instead of here?

File: src/main/java/frc/robot/RobotContainer.java [hunk 4]
```diff
@@ -165,7 +170,26 @@ public class RobotContainer {
 
     // SENDABLE CHOOSER
     private final SendableChooser<Command> chooser = new SendableChooser<>();
+    // private final Command chargeCommand = new ChargeCommand(swerveDrive, odometry, trajectories);
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/RobotContainer.java]: Looks like all of this is a duplicate

File: src/main/java/frc/robot/Constants.java [hunk 5]
```diff
@@ -21,9 +21,12 @@ public final static class DrivetrainConstants {
 
         public static final double DRIFT_DEADBAND = 0.1;
         public static final double ROTATION_DEADBAND = 0.006;
+        public static final double BUTTON_DEADBAND = 0.2;
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/Constants.java]: If you want to use a deadband (which I'm not totally sure is necessary), it should probably be much smaller than 0.2. I'd suggest somewhere closer to 0.01

File: src/main/java/frc/robot/RobotContainer.java [hunk 6]
```diff
@@ -276,13 +305,24 @@ private void configureButtonBindings() {
           },
           swerveDrive));
 
+
+      fieldCentricButton.onTrue(new InstantCommand(
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/RobotContainer.java]: Duplicate code

File: src/main/java/frc/robot/RobotContainer.java [hunk 7]
```diff
@@ -276,13 +305,24 @@ private void configureButtonBindings() {
...
       fieldCentricButton.onTrue(new InstantCommand(
           () -> {
             System.out.println("FIELD CENTRIC TOGGLED");
             swerveDrive.toggleFieldCentric();
           }, swerveDrive));
       
       balanceChargingStationButton.whileTrue(levelChargingStationCommand);
+      balanceChargingStationButton.whileTrue(new InstantCommand(
+          () -> {
+            ledSubsystem.balancingColor();;
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/RobotContainer.java]: Move this to the LevelChargingStationCommand, instead of attaching it to the button.

File: src/main/java/frc/robot/RobotContainer.java [hunk 8]
```diff
@@ -338,9 +378,13 @@ private void configureButtonBindings() {
...
       // alignLeftSubstationButton.whileTrue(alignLeftSubstation);
       // alignRightSubstationButton.whileTrue(alignRightSubstation);
+
+      wantCone.onTrue(new RunCommand(() -> ledSubsystem.coneColor()));
+      wantCube.onTrue(new RunCommand(() -> ledSubsystem.cubeColor()));
+
       
     }
-
+
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/RobotContainer.java]: Whitespace :(

File: src/main/java/frc/robot/RobotContainer.java [hunk 9]
```diff
@@ -383,6 +433,8 @@ public void periodic() {
 
       pitch.setDouble(odometry.getPitch().getDegrees());
       roll.setDouble(odometry.getRoll().getDegrees());
+
+      ledSubsystem.runLEDs();
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/RobotContainer.java]: This should be in the LEDSubsystem's periodic function, not here

File: src/main/java/frc/robot/RobotContainer.java [hunk 10]
```diff
@@ -396,4 +448,5 @@ public void buildAutoEventMap(){
       eventMap.put("telescopeOut",
         new RunCommand(() -> telescopeSubsystem.setTelescopePosition(ArmConstants.telescopeOuterSetpoint)));
     }
+
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/RobotContainer.java]: Whitespace :(

File: src/main/java/frc/robot/commands/LevelChargingStationCommand.java [hunk 11]
```diff
@@ -4,12 +4,13 @@
 import edu.wpi.first.wpilibj2.command.PIDCommand;
 import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
 import frc.robot.common.Odometry;
+import frc.robot.subsystems.LEDSubsystem;
 import frc.robot.subsystems.SwerveDriveSubsystem;
 
 public class LevelChargingStationCommand extends SequentialCommandGroup {
+
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/commands/LevelChargingStationCommand.java]: Whitespace :(

File: src/main/java/frc/robot/commands/LevelChargingStationCommand.java [hunk 12]
```diff
@@ -4,12 +4,13 @@
...
 import frc.robot.common.Odometry;
+import frc.robot.subsystems.LEDSubsystem;
 import frc.robot.subsystems.SwerveDriveSubsystem;
 
 public class LevelChargingStationCommand extends SequentialCommandGroup {
+    
 
-    public LevelChargingStationCommand(Odometry odometry, SwerveDriveSubsystem swerveDriveSubsystem) {
-
+    public LevelChargingStationCommand(Odometry odometry, SwerveDriveSubsystem swerveDriveSubsystem, LEDSubsystem ledSubsystem) {
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/commands/LevelChargingStationCommand.java]: You add the parameter, but do you ever actually use it here?

File: src/main/java/frc/robot/commands/autonomous/ChargeCommand.java [hunk 13]
```diff
@@ -43,4 +43,4 @@ public ChargeCommand(SwerveDriveSubsystem swerveDriveSubsystem, Odometry odometr
     // Use addRequirements() here to declare subsystem dependencies.
     addRequirements(swerveDriveSubsystem);
   }
-}
+}
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/commands/autonomous/ChargeCommand.java]: Whitespace :(

File: src/main/java/frc/robot/common/Odometry.java [hunk 14]
```diff
@@ -88,6 +88,7 @@ public void setAngleOffset(double offset) {
     }
 
     public Pose2d getPose() {
+        System.out.println(odometry.getPoseMeters());
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/common/Odometry.java]: ?

File: src/main/java/frc/robot/subsystems/LEDSubsystem.java [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+import java.lang.System.Logger.Level;
+
+import edu.wpi.first.wpilibj.DriverStation;
+import edu.wpi.first.wpilibj.DriverStation.Alliance;
+import edu.wpi.first.wpilibj.motorcontrol.Spark;
+import frc.robot.Constants.LEDConstants;
+import frc.robot.common.Odometry;
+import frc.robot.Constants.LEDColors;
+
+public class LEDSubsystem {
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/subsystems/LEDSubsystem.java]: This should have `extends SubsystemBase` after it. Look at other subsystem classes to see how they're built.

File: src/main/java/frc/robot/Constants.java [hunk 16]
```diff
@@ -96,6 +99,29 @@ public final static class DriverConstants {
         public final static double speedMultiplier = 3;
         public final static double angleMultiplier = 3;
     }
+    
+    public final static class LEDConstants {
+        public final static int BLINKIN_CHANNEL = 0;
+    }
+
+    public enum LEDColors {
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/Constants.java]: On reflection, calling this LEDColor would make more sense, as whenever you refer to the enum you're only doing so in reference to a single color

File: src/main/java/frc/robot/subsystems/LEDSubsystem.java [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+            ledColor = LEDColors.RED_ALLIANCE;
+    }
+
+    public void clawClosedColor() {
+        ledColor = LEDColors.CLAW_CLOSED;
+    }
+
+    public void balancingColor() {
+        if(odometry.getInclineAngle().getDegrees() > 0.05)
+        ledColor = LEDColors.BALANCING;
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/subsystems/LEDSubsystem.java]: Tab this line in

File: src/main/java/frc/robot/subsystems/LEDSubsystem.java [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+        else
+        ledColor = LEDColors.BALANCED;
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/subsystems/LEDSubsystem.java]: Tab this line in

File: src/main/java/frc/robot/subsystems/LEDSubsystem.java [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+    
+
+    public void startColor() {
+        if(DriverStation.getAlliance() == Alliance.Blue)
+            ledColor = LEDColors.BLUE_ALLIANCE;
+        else
+            ledColor = LEDColors.RED_ALLIANCE;
+    }
+
+    public void clawClosedColor() {
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/subsystems/LEDSubsystem.java]: Call all of these functions `set...Color` to reflect normal Java etiquette.

File: src/main/java/frc/robot/RobotContainer.java [hunk 20]
```diff
@@ -350,6 +394,12 @@ public Command getAutonomousCommand() {
       // An ExampleCommand will run in autonomous
       // odometry.zeroHeading();
       return chooser.getSelected();
+
+    }
+
+    public void startLEDs() {
+      ledSubsystem.startColor();
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/RobotContainer.java]: Instead of making a method in robotContainer which only calls a method in a different subsystem, make a getter method in the robotContainer that returns the ledSubsystem, and use that to call `startColor` directly

**Review by BrandonWerbel — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Constants.java (hunk 5 above)
Comment (BenBerol) [reply src/main/java/frc/robot/Constants.java]: If we are pressing the button it will almost certainly go above 0.2, and I am just building in a deadband so that barely pressing the button doesn't trigger the lights

File: src/main/java/frc/robot/RobotContainer.java (hunk 20 above)
Comment (BenBerol) [reply src/main/java/frc/robot/RobotContainer.java]: ? I don't understand. I can make a getter method but them how will I use that to start color?

File: src/main/java/frc/robot/RobotContainer.java (hunk 9 above)
Comment (BenBerol) [reply src/main/java/frc/robot/RobotContainer.java]: If I make a periodic function in ledsubsystem, will I have to call it in the teleop periodic in robot?

File: src/main/java/frc/robot/RobotContainer.java [hunk 21]
```diff
@@ -137,35 +138,43 @@ public class RobotContainer {
...
 
+
+    private LEDSubsystem ledSubsystem = new LEDSubsystem(odometry);
+
     private DoubleSolenoid leftSolenoid = new DoubleSolenoid(GrabberConstants.PNEUMATICS_HUB_ID, PneumaticsModuleType.REVPH, GrabberConstants.LEFT_SOLENOID_FORWARD, GrabberConstants.LEFT_SOLENOID_REVERSE);
     private DoubleSolenoid rightSolenoid = new DoubleSolenoid(GrabberConstants.PNEUMATICS_HUB_ID, PneumaticsModuleType.REVPH, GrabberConstants.RIGHT_SOLENOID_FORWARD, GrabberConstants.RIGHT_SOLENOID_REVERSE);
-    private GrabberSubsystem grabberSubsystem = new GrabberSubsystem(leftSolenoid, rightSolenoid);
-
+    private GrabberSubsystem grabberSubsystem = new GrabberSubsystem(leftSolenoid, rightSolenoid, ledSubsystem);
+
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/RobotContainer.java]: Whitespace :(

File: src/main/java/frc/robot/RobotContainer.java [hunk 22]
```diff
@@ -137,17 +138,25 @@ public class RobotContainer {
...
+    
     private TalonFX elevatorMotor = new TalonFX(ElevatorConstants.ELEVATOR_MOTOR_ID);
     private ElevatorSubsystem elevatorSubsystem = new ElevatorSubsystem(elevatorMotor);
 
+
     // LOAD TRAJECTORIES
     private final TrajectoryLoader trajectoryLoader = new TrajectoryLoader();
     private final HashMap<String, Trajectory> trajectories = trajectoryLoader.loadTrajectories();
 
+    // SENDABLE CHOOSER
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/RobotContainer.java]: Why did you move this here?

File: src/main/java/frc/robot/RobotContainer.java [hunk 23]
```diff
@@ -156,16 +165,16 @@ public class RobotContainer {
     // private final AlignByAprilTag alignLeftSubstation = new AlignByAprilTag(swerveDrive, limelight, odometry, -0.584, -1);
 
 
+
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/RobotContainer.java]: whitespace :(

File: src/main/java/frc/robot/RobotContainer.java [hunk 24]
```diff
@@ -156,16 +165,16 @@ public class RobotContainer {
...
     private final ChargeCommand chargeCommand = new ChargeCommand(swerveDrive, odometry, eventMap);
     private final LeftUpperConeAutonomous leftUpperConeAutonomous = new LeftUpperConeAutonomous(grabberSubsystem, pivotSubsystem, telescopeSubsystem, swerveDrive, odometry, eventMap);
     private final RightUpperConeAutonomous rightUpperConeAutonomous = new RightUpperConeAutonomous(grabberSubsystem, pivotSubsystem, telescopeSubsystem, swerveDrive, odometry, eventMap);
-    private final MiddleAutonomous middleAutonomous = new MiddleAutonomous(grabberSubsystem, pivotSubsystem, telescopeSubsystem, swerveDrive, odometry, eventMap);
+    private final MiddleAutonomous middleAutonomous = new MiddleAutonomous(grabberSubsystem, pivotSubsystem, telescopeSubsystem, swerveDrive, odometry, eventMap, ledSubsystem);
 
-    // SENDABLE CHOOSER
-    private final SendableChooser<Command> chooser = new SendableChooser<>();
 
+
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/RobotContainer.java]: whitespace :(

File: src/main/java/frc/robot/RobotContainer.java [hunk 25]
```diff
@@ -283,6 +298,11 @@ private void configureButtonBindings() {
           }, swerveDrive));
       
       balanceChargingStationButton.whileTrue(levelChargingStationCommand);
+      balanceChargingStationButton.whileTrue(new InstantCommand(
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/RobotContainer.java]: Move this `InstantCommand` to the LevelChargingStation `addCommands` function instead of attaching it to the button here

File: src/main/java/frc/robot/RobotContainer.java (hunk 9 above)
Comment (BrandonWerbel) [reply src/main/java/frc/robot/RobotContainer.java]: Nope! Look at `ExampleSubsystem` to see an example. The `@Override` tag tells the computer that it works with it's parent (`subsystemBase`) class, which runs multiple times a second

File: src/main/java/frc/robot/commands/LevelChargingStationCommand.java [hunk 26]
```diff
@@ -4,12 +4,12 @@
 import edu.wpi.first.wpilibj2.command.PIDCommand;
 import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
 import frc.robot.common.Odometry;
+import frc.robot.subsystems.LEDSubsystem;
 import frc.robot.subsystems.SwerveDriveSubsystem;
 
 public class LevelChargingStationCommand extends SequentialCommandGroup {
 
     public LevelChargingStationCommand(Odometry odometry, SwerveDriveSubsystem swerveDriveSubsystem) {
-
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/commands/LevelChargingStationCommand.java]: Whitespace :(

File: src/main/java/frc/robot/commands/autonomous/MiddleAutonomous.java [hunk 27]
```diff
@@ -47,4 +50,4 @@ public MiddleAutonomous(GrabberSubsystem grabberSubsystem, PivotSubsystem pivotS
         );
     }
     
-}
+}
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/commands/autonomous/MiddleAutonomous.java]: Whitespace :(

**Review by BrandonWerbel — CHANGES_REQUESTED** [review]

**Comment by BrandonWerbel** [comment]
@BenGitsMagic Getting there! Before you push your next changes, check the summary of changes to make sure nothing shows up as changed other than what you intended to change. Looking for those is where all my whitespace comments come from :)

File: src/main/java/frc/robot/subsystems/LEDSubsystem.java [hunk 28]
```diff
@@ -0,0 +1,61 @@
...
+
+    private Odometry odometry;
+
+    public LEDSubsystem(Odometry odometry) {
+        blinkin = new Spark(LEDConstants.BLINKIN_CHANNEL);
+        
+        this.odometry = odometry;
+    }
+
+    public void runLEDs() {
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/subsystems/LEDSubsystem.java]: Why is this its own method? It's only used once in the periodic method

File: src/main/java/frc/robot/commands/LevelChargingStationCommand.java [hunk 29]
```diff
@@ -1,14 +1,17 @@
...
 import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.wpilibj2.command.InstantCommand;
 import edu.wpi.first.wpilibj2.command.PIDCommand;
 import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
+import frc.robot.Constants.LimelightDirections;
 import frc.robot.common.Odometry;
+import frc.robot.subsystems.LEDSubsystem;
 import frc.robot.subsystems.SwerveDriveSubsystem;
 
 public class LevelChargingStationCommand extends SequentialCommandGroup {
```
Comment (BrandonWerbel) [inline src/main/java/frc/robot/commands/LevelChargingStationCommand.java]: Because this is a `SequentialCommandGroup`, it runs commands one at a time, in the same order as listed in `addCommands`. Since you want to be running the `balanceRobot` command *at the same time* as the LEDs change, you need to make the command a `ParallelCommandGroup`, which runs all the commands in `addCommands` at the same time.

**Review by BrandonWerbel — CHANGES_REQUESTED** [review]
